### PR TITLE
Added telemetry to DevServer

### PR DESCRIPTION
**Pull request by carldebilly**
Closes https://github.com/unoplatform/uno-private/issues/1373

# ✨ Feature

This PR adds comprehensive telemetry support to the DevServer and related components, updates dependencies for telemetry, and introduces end-to-end tests validating telemetry flows.

- Integrate Uno.DevTools.Telemetry across server, host, and discovery pipelines
- Update project files to register telemetry services and adapt SLN for new test projects
- Add extensive MSTest test suites for telemetry, scoped services, and helper classes

## What is the new behavior? 🚀
This pull request introduces telemetry enhancements, dependency updates, and new tests for the Uno DevServer. The changes focus on improving telemetry data handling, adding new test coverage, and updating project dependencies to support testing and telemetry features.

### Telemetry Enhancements

* Added a detailed telemetry events inventory to `Telemetry/Telemetry.md`, documenting event properties, measurements, and anonymization practices for GDPR/privacy compliance.

### Dependency Updates

* Updated `global.json` to include the `MSTest.Sdk` dependency (version 3.8.3) for enhanced testing capabilities.

#### DevServer Tests
* Implemented multiple tests in `DevServerTests.cs` to validate DevServer functionality, including startup, output capture, and clean shutdown.
* Added global usings for common libraries (`FluentAssertions`, `Microsoft.Extensions.Logging`, etc.) in `GlobalUsings.cs` to simplify test code.

#### Telemetry Tests
* Created `TelemetryHashHelperTests.cs` to ensure deterministic and distinct hashing for telemetry data, handling edge cases like null and empty inputs.
* Added `Telemetry/ContextualTelemetryTests.cs` to validate telemetry behavior in global-only scenarios.
* Added `Telemetry/ServerTelemetryTests.cs` to ensure server logs connection events correctly during telemetry cycles.

#### Scoped Service Isolation
* Developed `ScopedServiceIsolationTests.cs` to verify scoped service isolation between connections and proper telemetry redirection.

#### Helper Classes
* Added `SolutionHelper.cs` to facilitate temporary solution file creation for testing purposes.

## PR Checklist ✅

Please check if your PR fulfills the following requirements:

- [X] 📝 Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.
- [X] 🧪 Added [Runtime tests, UI tests, or a manual test sample](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] 📚 Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] 🖼️ Validated PR `Screenshots Compare Test Run` results.
- [X] ❗ Contains **NO** breaking changes

### CODE REVIEW

File: src/Uno.UI.RemoteControl.Host/Program.cs [hunk 1]
```diff
@@ -1,112 +1,265 @@
...
-			// Once the app has started, we use the logger from the host
-			Uno.Extensions.LogExtensionPoint.AmbientLoggerFactory = host.Services.GetRequiredService<ILoggerFactory>();
+				// STEP 3: Use global telemetry for server-wide events
+				// Track devserver startup using global telemetry service
+				telemetry = globalServiceProvider.GetService<ITelemetry>();
+				var startupProperties = new Dictionary<string, string>
+				{
+					["HasSolution"] = (solution != null).ToString(),
+					["MachineName"] = TelemetryHashHelper.Hash(Environment.MachineName),
+					["OSVersion"] = Environment.OSVersion.ToString(),
```
Comment (carldebilly) [inline src/Uno.UI.RemoteControl.Host/Program.cs]: Probably already in the telemetry service, would be duplicate information.

File: Telemetry/Telemetry.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+| Event Name                      | Properties (string)                                                                                      | Measurements (double)                        | Sensitive / Notes                            | Scope         |
+|----------------------------------|---------------------------------------------------------------------------------------------------------|----------------------------------------------|----------------------------------------------|---------------|
+| **DevServer.Startup**            | HasSolution, MachineName🔴, OSVersion                                                                   | ProcessorCount                              | MachineName🔴 is anonymized; OSVersion is raw | Global        |
+| **DevServer.Shutdown**           | ShutdownType ("Graceful"/"Crash")                                                                      | UptimeSeconds                               |                                              | Global        |
```
Comment (carldebilly) [inline Telemetry/Telemetry.md]: "Crash" is a little too intense for now, should be "forced" or "killed"

File: src/Uno.UI.RemoteControl.Host/Extensibility/AddIns.cs [hunk 3]
```diff
@@ -8,90 +9,151 @@
...
-				if (result.error is { Length: > 0 })
-				{
-					_log.Log(LogLevel.Warning, new Exception(result.error), msg + " (cf. inner exception for more details.)");
-				}
-				else
-				{
-					result = ProcessHelper.RunProcess("dotnet", DumpTFM("diagnostic"), wd);
+			// Note: We include the targets "on the fly" so if a project uses Microsoft.NET.Sdk instead of Uno.Sdk, we will still have the targets defined.
+			var targetsFile = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!,
+				"DevServer.Custom.Targets");
```
Comment (dr1rrb) [inline src/Uno.UI.RemoteControl.Host/Extensibility/AddIns.cs]: What a weird indentation ;)

File: src/Uno.UI.RemoteControl.Host/Extensibility/AddInsExtensions.cs [hunk 4]
```diff
@@ -1,17 +1,27 @@
...
 
 public static class AddInsExtensions
 {
 	public static IWebHostBuilder ConfigureAddIns(this IWebHostBuilder builder, string solutionFile)
 	{
-		AssemblyHelper.Load(AddIns.Discover(solutionFile), throwIfLoadFailed: false);
+		return builder.ConfigureServices(services =>
+		{
+			// Get telemetry service for add-in discovery and loading
+			var serviceProvider = services.BuildServiceProvider();
```
Comment (dr1rrb) [inline src/Uno.UI.RemoteControl.Host/Extensibility/AddInsExtensions.cs]: Is this a recommended pattern to use MS DI? Sounds weird to me to resolve a service during the registration phase.

File: src/Uno.UI.RemoteControl.Host/Helpers/AssemblyHelper.cs [hunk 5]
```diff
@@ -1,38 +1,92 @@
...
 
-		return assemblies.ToImmutable();
+			var errorMeasurements = new Dictionary<string, double>
+			{
+				["DurationMs"] = Stopwatch.GetElapsedTime(startTime).TotalMilliseconds,
+				["LoadedAssemblies"] = loadedCount,
+				["FailedAssemblies"] = failedCount,
+			};
+
+			telemetry?.TrackEvent("AddIn.Loading.Error", errorProperties, errorMeasurements);
```
Comment (dr1rrb) [inline src/Uno.UI.RemoteControl.Host/Helpers/AssemblyHelper.cs:80]: Not a fan to track an "AddIn" event in class named "AssemblyHelper" ;) I would track event from the caller instead.

File: src/Uno.UI.RemoteControl.Host/Extensibility/AddIns.cs [hunk 6]
```diff
@@ -8,90 +9,151 @@
...
 
-	public static IImmutableList<string> Discover(string solutionFile)
+	public static IImmutableList<string> Discover(string solutionFile, ITelemetry? telemetry = null)
 	{
-		// Note: We include the targets "on the fly" so if a project uses Microsoft.NET.Sdk instead of Uno.Sdk, we will still have the targets defined.
-		var targetsFile = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!, "DevServer.Custom.Targets");
+		var startTime = Stopwatch.GetTimestamp();
+		var discoveryProperties = new Dictionary<string, string>
+		{
+			["SolutionId"] = TelemetryHashHelper.Hash(solutionFile),
```
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl.Host/Extensibility/AddIns.cs]: There's already a property for this, CurrentHashPath, which we should use instead of adding another one

File: src/Uno.UI.RemoteControl.Host/Helpers/AssemblyHelper.cs [hunk 7]
```diff
@@ -1,38 +1,92 @@
...
+				["ErrorMessage"] = ex.Message,
+				["ErrorType"] = ex.GetType().Name,
+			};
 
-		return assemblies.ToImmutable();
+			var errorMeasurements = new Dictionary<string, double>
+			{
+				["DurationMs"] = Stopwatch.GetElapsedTime(startTime).TotalMilliseconds,
+				["LoadedAssemblies"] = loadedCount,
+				["FailedAssemblies"] = failedCount,
```
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl.Host/Helpers/AssemblyHelper.cs]: this is going to be too large, we probably should only be listing the failed assemblies

**Review by dr1rrb — APPROVED** [review]

File: src/Uno.UI.RemoteControl.Host/Program.cs [hunk 8]
```diff
@@ -1,112 +1,255 @@
...
-			var httpPort = 0;
-			var parentPID = 0;
-			var solution = default(string);
+			var startTime = Stopwatch.GetTimestamp();
+
+			// Set up graceful shutdown handling
+			using var cancellationTokenSource = new CancellationTokenSource();
+			var shutdownRequested = false;
+
+			Console.CancelKeyPress += (sender, e) =>
```
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl.Host/Program.cs:34]: This might not work on all platforms because of non-interactive shells

File: src/Uno.UI.RemoteControl.Host/Program.cs [hunk 9]
```diff
@@ -1,112 +1,255 @@
...
-			// Once the app has started, we use the logger from the host
-			Uno.Extensions.LogExtensionPoint.AmbientLoggerFactory = host.Services.GetRequiredService<ILoggerFactory>();
+				host.Services.GetService<IIdeChannel>();
+
+				// STEP 3: Use global telemetry for server-wide events
+				// Track devserver startup using global telemetry service
+				var startupProperties = new Dictionary<string, string>
+				{
+					["HasSolution"] = (solution != null).ToString(),
+					["MachineName"] = TelemetryHashHelper.Hash(Environment.MachineName),
```
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl.Host/Program.cs]: This is already included in the default telemetry properties

File: src/Uno.UI.RemoteControl.Host/RemoteControlServer.cs [hunk 10]
```diff
@@ -297,12 +310,26 @@ private async Task ProcessPingFrame(Frame frame)
...
 		{
 			var msg = JsonConvert.DeserializeObject<ProcessorsDiscovery>(frame.Content)!;
 			var serverAssemblyName = typeof(IServerProcessor).Assembly.GetName().Name;
 
+			// Track processor discovery start
+			var discoveryProperties = new Dictionary<string, string>
+			{
+				["AppInstanceId"] = TelemetryHashHelper.Hash(msg.AppInstanceId),
+				["BasePath"] = TelemetryHashHelper.Hash(msg.BasePath),
+				["IsFile"] = File.Exists(msg.BasePath).ToString()
```
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl.Host/RemoteControlServer.cs]: This is too generic, let's prefix it.

File: src/Uno.UI.RemoteControl.Server/Helpers/TelemetryHashHelper.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
+using System;
+using System.Runtime.CompilerServices;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Uno.UI.RemoteControl.Server.Helpers
+{
+	public static class TelemetryHashHelper
+	{
+		public static string Hash(object? input)
```
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl.Server/Helpers/TelemetryHashHelper.cs]: Let's use the same that the devtools library has

File: src/Uno.UI.RemoteControl.Host/Extensibility/AddIns.cs (hunk 3 above)
Comment (carldebilly) [reply src/Uno.UI.RemoteControl.Host/Extensibility/AddIns.cs]: Resharper reformatting... the indentation is correct, it's just how I'm doing it usually.

Comment (carldebilly) [reply src/Uno.UI.RemoteControl.Host/Extensibility/AddIns.cs]: Will be part of my next push.

File: src/Uno.UI.RemoteControl.Host/Extensibility/AddInsExtensions.cs (hunk 4 above)
Comment (carldebilly) [reply src/Uno.UI.RemoteControl.Host/Extensibility/AddInsExtensions.cs]: Definitively not a recommended pattern :-)

Even if in this case it wasn't a problem because there was no possibility of _leaks_ in the services, that's definitively not a good inspiration for others. Will be fixed in my next push.

File: src/Uno.UI.RemoteControl.Host/Program.cs (hunk 8 above)
Comment (carldebilly) [reply src/Uno.UI.RemoteControl.Host/Program.cs:34]: That's a best-effort and there's a fallback for automations. It works for unit tests and that's enough for me.
